### PR TITLE
Issue 11: Game Engine

### PR DESCRIPTION
**Pull request by Hyunse**
#11 

1. Game.js
Just one game engine class

2. Words.js
Words Array for playing a game

### CODE REVIEW

File: server/engine/Game.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  /**
+   * @param {string} id : Room Name 
+   * @param {array} players : game players array
+   */
+  constructor(id, players) {
+    this.roomName = id;
+    this.guesser = '';
+    this.word = '';
+    this.point = 0;
```
Comment (InsafKhamzin) [inline server/engine/Game.js]: Each player should have points,  not the game

File: server/engine/Game.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    this.round = 0;
+    this.wordArray = [];
+    this.players = players;
```
Comment (InsafKhamzin) [inline server/engine/Game.js]: Maybe it is better to have separate Player class with **id** and **isGuesser** properties

File: server/engine/Game.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+class Game {
+  ADD_POINT = 100;
+
+  /**
+   * @param {string} id : Room Name 
+   * @param {array} players : game players array
+   */
+  constructor(id, players) {
+    this.roomName = id;
```
Comment (InsafKhamzin) [inline server/engine/Game.js]: You don't have to store roomName here. We will have **roomName** with a **Game** object in our key-value cache.

File: server/engine/Game.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  /**
+   * Check Guesser's Answer
+   * @param {string} answer 
+   */
+  checkAnswer(answer) {
+    const word = this.word;
+    let correct = false;
+
+    if (word === answer) {
+      this.point += this.ADD_POINT;
```
Comment (InsafKhamzin) [inline server/engine/Game.js]: Need to add points to each user

**Review by InsafKhamzin — COMMENTED** [review]
Nice PR! Almost ready game engine, just need to add some little changes.

File: server/engine/Game.js (hunk 2 above)
Comment (darrenMabbayad) [reply server/engine/Game.js]: I agree with @InsafKhamzin. Should make it easier to determine how points are distributed to each player.

Comment (Hyunse) [reply server/engine/Game.js]: @InsafKhamzin @darrenMabbayad I added Player.js and changed that point will be calculated individually.

**Review by InsafKhamzin — APPROVED** [review]
Looks good to me! Things to discuss: maybe you don't need isGiver property. You can just pass an array of playerId and clues to givePoints() and figure out who is gonna get points or not.

**Review by darrenMabbayad — APPROVED** [review]
Everything looks good to me! We can expand on this later on to make things more complicated if we quickly get a full demo working.

File: server/engine/Game.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+       player.isGiver = false;
+    });
+  }
+
+  /**
+   * Check Guesser's Answer
+   * @param {string} answer
+   */
+  checkAnswer(answer) {
+    return this.word === answer ? true : false;
```
Comment (aeciorc) [inline server/engine/Game.js]:
```suggestion
    return this.word === answer;
```

File: server/engine/Game.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  /**
+   * Get Next Guesser
+   * @param {number} nextRound
+   */
+  setNextGuesser(nextRound) {
+    if (nextRound >= this.players.length) {
+      this.players[nextRound - this.players.length].isGuesser = true;
+    } else {
+      this.players[nextRound].isGuesser = true;
+    }
```
Comment (aeciorc) [inline server/engine/Game.js]:
```suggestion
   this.players[nextRound%this.player.length] = true;
```

File: server/engine/Game.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+  /**
+   * Give Point when answer is correct
+   */
+  givePoints() {
+    this.players.map((player) => {
+      // Guesser gets 200 points
+      if (player.isGuesser === true) player.addPoint(this.GUESS_POINT);
+
+      // Givers who gives not duplicated words get 100 points
+      if (player.isGiver === true) player.addPoint(this.CLUE_POINT);
```
Comment (aeciorc) [inline server/engine/Game.js]:
```suggestion
      if (player.isGuesser) player.addPoint(this.GUESS_POINT);

      // Givers who gives not duplicated words get 100 points
      if (player.isGiver) player.addPoint(this.CLUE_POINT);
```

File: server/engine/Game.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    return this.word === answer ? true : false;
+  }
+
+  /**
+   * Go to Next Round
+   */
+  nextRound() {
+    this.round = this.round + 1;
+    this.initPlayerRoles();
+    this.setNextGuesser(this.round);
```
Comment (aeciorc) [inline server/engine/Game.js:45]: What about the giver roles?

Comment (Hyunse) [reply server/engine/Game.js:45]: Player who give a clue to a guesser. If other players' clues are duplicated, they can't be a giver. If answer is correct, only a guesser and a giver can get points in a round.

**Review by aeciorc — APPROVED** [review]

**Pull request merged**